File: Problems/No.1472_Design_Browser_History/design_browser_history.py
class BrowserHistory(object):

    def __init__(self, homepage):
        """
        :type homepage: str
        """
        self.DBH = [homepage]
        self.curPid = 0
        self.lastPid = 0
        

    def visit(self, url):
        """
        :type url: str
        :rtype: None
        """
        self.curPid += 1
        if (len(self.DBH) > self.curPid):
            self.DBH[self.curPid] = url
        else:
            self.DBH.append(url)
        self.lastPid = self.curPid
        

    def back(self, steps):
        """
        :type steps: int
        :rtype: str
        """
        self.curPid = max(0, self.curPid-steps)
        return self.DBH[self.curPid]


    def forward(self, steps):
        """
        :type steps: int
        :rtype: str
        """
        self.curPid = min(self.lastPid, self.curPid+steps)
        return self.DBH[self.curPid]
        


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)

File: Problems/No.1472_Design_Browser_History/design_browser_history.cpp
class BrowserHistory {
private:
    std::vector<std::string> DBH;
    int curPid, lastPid;

public:
    BrowserHistory(string homepage) {

        DBH.push_back(homepage);
        curPid = 0;
        lastPid = 0;

    }
    
    void visit(string url) {

        curPid++;
        if (DBH.size() > curPid) DBH[curPid] = url;
        else DBH.push_back(url);
        lastPid = curPid;

    }
    
    string back(int steps) {
        
        curPid = std::max(0, curPid-steps);
        return DBH[curPid];

    }
    
    string forward(int steps) {

        curPid = std::min(lastPid, curPid+steps);
        return DBH[curPid];
        
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */
